### ISSUE TITLE
Type guard fails to narrow parameter of `any` type - v1.4

### ISSUE THREAD

**Issue by sccolbert** [Microsoft/TypeScript#1745]
If the following snippet, `dispatch` does not report an error, while `dispatch2` behaves as expected. From my reading of the spec section 4.20, both constructs should generate an error, but the first version fails to narrow the type of `object` from `any` to `string`. This also occurs with the `instanceof` operator.

```typescript
class Foo {
    a: number;
}


function doStuff(foo: Foo): void { }


function dispatch(object: any): void {
    if (typeof object === 'string') {
        doStuff(object);  // should fail but doesn't
    }
}

function dispatch2(object: number | string): void {
    if (typeof object === 'string') {
        doStuff(object);  // fails as expected
    }
}
```

Example from spec:
```typescript
// In the example
function isLongString(obj: any) {
    return typeof obj === "string" && obj.length > 100;
}
// the 'obj' parameter has type string in the right operand of the && operator.
```

**Comment by danquirk** [comment]
Both calls to ```doStuff``` correctly report errors in the latest bits in master.
